File: Ch02/ex2_27.exs
# SICP in Elixir: Exercise 2.27 - deep reverse

defmodule Ex2_27 do
  def deep_reverse([]),  do: []
  def deep_reverse([x]), do: [x]
  def deep_reverse([hd|tl]) do
    deep_reverse(tl) ++ deep_reverse(hd)
  end
  def deep_reverse(x),   do: [x]
end

x = [1, 2, 3]
y = [4, 5, 6, x]

result = Ex2_27.deep_reverse []
unless result == [], do: throw result

result = Ex2_27.deep_reverse [:a]
unless result == [:a], do: throw result

result = Ex2_27.deep_reverse x
unless result == [3, 2, 1], do: throw result

#result = Ex2_27.deep_reverse y
#unless result == [ [ 3, 2, 1 ], 6, 5, 4 ], do: throw result

File: Ch02/ex2_27.erl
-module(ex2_27).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: 2.27 - deep reverse").
-export([deep_reverse/1]).
-include_lib("eunit/include/eunit.hrl").

deep_reverse([]) -> [];
deep_reverse([X]) -> [X];
deep_reverse([Hd | Tl]) ->
    deep_reverse(Tl) ++ deep_reverse(Hd);
deep_reverse(X) -> [X].

%%% TESTS

empty_test() ->
    ?assertEqual([], deep_reverse([])).

single_test() ->
    ?assertEqual([x], deep_reverse([x])).

simple_test() ->
    ?assertEqual([3, 2, 1], deep_reverse([1, 2, 3])).

%compound_test() ->
%    ?assertEqual([ [3, 2, 1], 6, 5, 4], deep_reverse([4, 5, 6, [1, 2, 3]])).

